# Домашнее задание к лекции "Функции и работа с данными"

(Выполнил: Черевко Владимир)

In [1]:
import pandas as pd

## Задание 1

In [3]:
ratings_df = pd.read_csv('ratings.csv')
#ratings_df.head()
ratings_df["rating"].unique()

array([2.5, 3. , 2. , 4. , 3.5, 1. , 5. , 4.5, 1.5, 0.5])

Изсходя из имеющихся значений столбца rating, сделаем функцию со следующей логикой.

In [6]:
def rate_class(rating):
    if rating <= 2.:
        return "низкий рейтинг"
    elif rating <= 4.:
        return "средний рейтинг"
    elif rating in [4.5, 5.]:
        return "высокий рейтинг"

In [11]:
ratings_df["class"] = ratings_df["rating"].apply(rate_class)
#ratings_df["class"].unique()
ratings_df

userId  movieId  rating   timestamp            class
0            1       31     2.5  1260759144  средний рейтинг
1            1     1029     3.0  1260759179  средний рейтинг
2            1     1061     3.0  1260759182  средний рейтинг
3            1     1129     2.0  1260759185   низкий рейтинг
4            1     1172     4.0  1260759205  средний рейтинг
...        ...      ...     ...         ...              ...
99999      671     6268     2.5  1065579370  средний рейтинг
100000     671     6269     4.0  1065149201  средний рейтинг
100001     671     6365     4.0  1070940363  средний рейтинг
100002     671     6385     2.5  1070979663  средний рейтинг
100003     671     6565     3.5  1074784724  средний рейтинг

[100004 rows x 5 columns]

## Задание 2

In [30]:
keywords_df = pd.read_csv('keywords.csv')
keywords_df.head(20)
keywords_df["keyword"].unique()[:20]

array(['вк', 'одноклассники', 'порно', 'ютуб', 'вконтакте',
       'одноклассники моя страница', 'майл', 'авито', 'переводчик',
       'яндекс', 'сбербанк онлайн', 'mail', 'игры', 'youtube',
       'вконтакте моя страница', 'секс', 'гугл', 'dr', 'алиэкспресс',
       'рамблер'], dtype=object)

In [31]:
def get_region(keyword):
    low_keyword = keyword.lower()
    
    geo_data = {
    'Центр': ['москва', 'тула', 'ярославль'],
    'Северо-Запад': ['петербург', 'псков', 'мурманск'],
    'Дальний Восток': ['владивосток', 'сахалин', 'хабаровск']
    }
    
    for key, value in geo_data.items():
        for word in value:
            if word in low_keyword:
                return key
    return "undefined"    

In [32]:
keywords_df["region"] = keywords_df["keyword"].apply(get_region)
keywords_df["region"].unique()

array(['undefined', 'Центр', 'Северо-Запад', 'Дальний Восток'],
      dtype=object)

## Задание 3

In [33]:
movies_df = pd.read_csv('movies.csv')
movies_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [34]:
ratings_df = pd.read_csv('ratings.csv')
ratings_df.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [54]:
year_lst = range(1950, 2011)

In [55]:
def production_year(title, year_lst):
    for year in year_lst:
        if str(year) in title:
            return year
    return 1900

In [56]:
movies_df["year"] = movies_df["title"].apply(production_year, year_lst=year_lst)

In [58]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9125 entries, 0 to 9124
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9125 non-null   int64 
 1   title    9125 non-null   object
 2   genres   9125 non-null   object
 3   year     9125 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 285.3+ KB


In [59]:
len(movies_df.movieId.unique())

9125

In [61]:
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100004 entries, 0 to 100003
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100004 non-null  int64  
 1   movieId    100004 non-null  int64  
 2   rating     100004 non-null  float64
 3   timestamp  100004 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [62]:
len(ratings_df.movieId.unique())

9066

Судя по информации выше, у нас в таблице с рейтингом присутствует меньше уникальных movieId (9066), чем в таблице с фильмами (там они все уникальны и их 9125). Получается, для части фильмов рейтинка нет. Предлагаю сделать связь таблиц по типу inner join, чтобы мы могле делать анализ только для тех фильмов, для которых рейтинг есть, иначе, наш анализ будет неинформативен.

In [66]:
movie_rating_df = movies_df.merge(ratings_df, on="movieId", how="inner")
movie_rating_df.head()
#len(movie_rating_df.movieId.unique()) //Мы получили 9066 фильмов, как и хотели

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   year  userId  rating   timestamp  
0  1995       7     3.0   851866703  
1  1995       9     4.0   938629179  
2  1995      13     5.0  1331380058  
3  1995      15     2.0   997938310  
4  1995      19     3.0   855190091

In [76]:
movie_rating_df.groupby("year")["rating"].mean().sort_values(ascending=False)

year
1957    4.014241
1972    4.011136
1952    4.000000
1954    3.994220
1951    3.983539
          ...   
2005    3.448434
2003    3.444777
1996    3.426600
1997    3.415764
2000    3.356104
Name: rating, Length: 62, dtype: float64

Из результата можно видеть, что в реальности, самый высокий рейтинг действительно имеют фильмы ранних годов, а фильмы поздних готов, как правило имеют средний рейтинг ниже. Но, при этом по годам распределение неравномерное.